ompi/mca/coll/base/coll_base_util.c

```diff
@@ -76,6 +76,12 @@ int ompi_coll_base_sendrecv_actual( const void* sendbuf, size_t scount,
     if (MPI_STATUS_IGNORE != status) {
         status->MPI_ERROR = err;
     }
+#if OPAL_ENABLE_FT_MPI
+    /* need to complete the recv in any case; otherwise the recv buffer is
+     * not safe to reuse */
+    ompi_request_cancel(req);
```

Comment:
Non FT errors may not be cancellable and could deadlock. What we need to do instead is verify if the error type is  cancellable, or free the request otherwise.